### PR TITLE
fix: extends `ElementHandle` to `Node`s

### PR DESCRIPTION
**Pull request by jrandolf-google**
This PR adjusts some incorrect types within Puppeteer:

- The base type parameter for `ElementHandle` is now `Node` instead of `Element`.
- The default type parameter for `ElementHandle` is still `Element`.
- `QueryHandler`s now use `Node` types.

### CODE REVIEW

**Comment by OrKoN** [comment]
In what cases can an ElementHandle contain a Node?

**Comment by jrandolf-google** [comment]
> In what cases can an ElementHandle contain a Node?

See `_createJSHandle`. Basically, ElementHandles are created from `Node`s (not `Element`s). For example, `evaluateHandle(() => document)` will return a valid `ElementHandle`. Similarly for queries involving `ShadowRoot`s.

**Comment by OrKoN** [comment]
are any updates needed for api.md for these changes?

**Comment by jrandolf-google** [comment]
> are any updates needed for api.md for these changes?

There are a lot of updates for `api.md`, but DocLint doesn't support the advanced typings made in previous stages. I propose we merge this and hold on release until we get some documentation (should be complete by next week).

**Comment by OrKoN** [comment]
@jrandolf wdyt about doing a release now (I assume previous PRs don't need api.md changes) and postponing landing this until the documentation is ready? Today or tomorrow we also need to roll M104 which would be nice to release as soon as possible too so I would not like to block a release on anything.

File: test/src/elementhandle.spec.ts [hunk 1]
```diff
@@ -572,10 +615,16 @@ describe('ElementHandle specs', function () {
...
+            return element.querySelector(`.${selector}`);
+          }
+          return null;
         },
         queryAll: (element, selector) => {
-          return element.querySelectorAll(`.${selector}`);
+          if (element instanceof Element || element instanceof Document) {
+            return element.querySelectorAll(`.${selector}`);
+          }
+          return [];
```
Comment (OrKoN) [inline test/src/elementhandle.spec.ts]: let's throw a runtime error when the element type is not right?

**Comment by OrKoN** [comment]
Let's also include the before/after examples in the docs.

File: test-d/ElementHandle.test-d.ts [hunk 2]
```diff
@@ -119,7 +119,7 @@ declare const handle: ElementHandle;
     await handle.$$eval(
       'some-element',
       (elements, value) => {
-        expectType<Element[]>(elements);
+        expectType<Node[]>(elements);
```
Comment (OrKoN) [inline test-d/ElementHandle.test-d.ts]: should it remain the Element still?

**Review by OrKoN — APPROVED** [review]

**Comment by Qlub53** [comment]
I used to use [page.$x(expression)](https://github.com/puppeteer/puppeteer/blob/v13.4.0/docs/api.md#pagexexpression) and 
[elementHandle.click([options])](https://github.com/puppeteer/puppeteer/blob/v13.4.0/docs/api.md#elementhandleclickoptions) with puppeteer v13.4 just fine.
```ts
import puppeteer from 'puppeteer';

const browser = await puppeteer.launch();
const page = await browser.newPage();
await page.goto('https://example.com');
const handles = await page.$x('xpath');
await handles[0].click();
```

But now since API signatures have changed, using [$x(expression)](https://github.com/puppeteer/puppeteer/blob/v16.2.0/docs/api/puppeteer.page._x.md) and [click(this, options)](https://github.com/puppeteer/puppeteer/blob/v16.2.0/docs/api/puppeteer.elementhandle.click.md) from v16.2 ends up with the following `tsc` compilation error:
```
error TS2684: The 'this' context of type 'ElementHandle<Node>' is not assignable to method's 'this' of type 'ElementHandle<Element>'.

7 await handles[0].click();
        ~~~~~~~~~~
```

How shall I transit from using `Element` to using `Node`?

**Comment by Qlub53** [comment]
I was able to get away with this problem by type casting.
```ts
await (handles[0] as ElementHandle<Element>).click();
```
Turns out it in fact still works at runtime and only bugs me during compile time.

It's not pretty, but it works. :/

**Pull request merged**